tests/test_batch_run.py

```diff
@@ -0,0 +1,125 @@
+        {"a": 3, "b": 2},
+    ]
+    assert _make_model_kwargs({"a": range(2), "b": range(2)}) == [
+        {"a": 0, "b": 0},
+        {"a": 0, "b": 1},
+        {"a": 1, "b": 0},
+        {"a": 1, "b": 1},
+    ]
+
+
+class MockAgent(Agent):
+    """
+    Minimalistic agent implementation for testing purposes
+    """
+
+    def __init__(self, unique_id, model, val):
+        super().__init__(unique_id, model)
+        self.unique_id = unique_id
+        self.val = val
+        self.local = 0
+
+    def step(self):
+        self.val += 1
+        self.local += 0.25
+
+
+class MockModel(Model):
+    """
+    Minimalistic model for testing purposes
+    """
+
+    def __init__(
+        self,
+        variable_model_param=None,
+        variable_agent_param=None,
+        fixed_model_param=None,
+        schedule=None,
+        **kwargs
+    ):
+        super().__init__()
+        self.schedule = BaseScheduler(self) if schedule is None else schedule
+        self.variable_model_param = variable_model_param
+        self.variable_agent_param = variable_agent_param
+        self.fixed_model_param = fixed_model_param
+        self.n_agents = 3
+        self.datacollector = DataCollector(
+            model_reporters={"reported_model_param": self.get_local_model_param},
+            agent_reporters={"agent_id": "unique_id", "agent_local": "local"},
+        )
+        self.running = True
+        self.init_agents()
+
+    def init_agents(self):
+        if self.variable_agent_param is None:
+            agent_val = 1
+        else:
+            agent_val = self.variable_agent_param
+        for i in range(self.n_agents):
+            self.schedule.add(MockAgent(i, self, agent_val))
+
+    def get_local_model_param(self):
+        return 42
+
+    def step(self):
+        self.datacollector.collect(self)
+        self.schedule.step()
+
+
+def test_batch_run():
+    result = batch_run(MockModel, {})
+    assert result == [
+        {
+            "RunId": 0,
+            "iteration": 0,
+            "Step": 1000,
+            "reported_model_param": 42,
+            "AgentID": 0,
+            "agent_id": 0,
+            "agent_local": 250.0,
+        },
+        {
+            "RunId": 0,
+            "iteration": 0,
+            "Step": 1000,
+            "reported_model_param": 42,
+            "AgentID": 1,
+            "agent_id": 1,
+            "agent_local": 250.0,
+        },
+        {
+            "RunId": 0,
+            "iteration": 0,
+            "Step": 1000,
+            "reported_model_param": 42,
+            "AgentID": 2,
+            "agent_id": 2,
+            "agent_local": 250.0,
+        },
+    ]
+
+
+def test_batch_run_with_params():
+    batch_run(
+        MockModel,
+        {
+            "variable_model_params": range(5),
+            "variable_agent_params": ["H", "E", "L", "L", "O"],
+        },
+    )
+
+
+def test_batch_run_single_core():
+    batch_run(MockModel, {}, nr_processes=1, iterations=10)
```

Comment:
To run Pools with Windows you have to use freeze_support otherwise it fails while also working all other OSs.
I thought we added Mac OS and Windows OS to our testing if not then this is just a good to know.  

```
if __name__ == "__main__":
    freeze_support()
```

Reply:
Hmm I was developing this on a windows machine and never had any issues. But from what I heard it should not hurt to add freeze_support!? This is only for running the tests, right?

Reply:
That is weird until I added freeze_support I was getting errors and saying use freeze_support and I have never had a multiprocessing batchrunner work unless I have used freeze_support. I am intrigued, let me dig a little deeper and see if I can find a reason.

Reply:
@Corvince as usual I learn a ton from reading your code. So aplogies it took so long, but thank you!

Apparently it all does boil down to pytest and how it is handling it tests although I could not find an explicit discussion of pytest and multiprocessing.

Regardless, the code is all good and just for reference I found this description of what was happening when I was using unittest and getting a runtime error: [differences between multiprocessing windows and linux](https://www.pythonforthelab.com/blog/differences-between-multiprocessing-windows-and-linux/l)